### ISSUE TITLE
Automatic Garagaopener

### ISSUE THREAD

**Issue by torsteinelv** [valentinfrlch/ha-llmvision#120]
Hello,

Sharing thoughts on a automatic garage opener automation, it works almost perferct. Only thing is that I do not understand how we can use the difference between the 8 frames that are captured and compare that the care actualy is driving in reverse since i dont want to port to open if the car i stand still parked outside the garage. Guess we can pull the sensor from the tesla integration but I wanted to see if it was possible with llmvision app only .

Sorry if this is created in incorrect place as there was no option for discussion.

`alias: Frigate - Car Detection with LLMVision
description: ""
triggers:
  - type: occupied
    device_id: a5cc13a0651766bdac964cf6106d6c71
    entity_id: f1e5af7728b6fb11f897a400d78cf4d4
    domain: binary_sensor
    trigger: device
    for:
      hours: 0
      minutes: 0
      seconds: 1
conditions: []
actions:
  - action: llmvision.stream_analyzer
    metadata: {}
    data:
      remember: true
      duration: 8
      max_frames: 8
      include_filename: false
      target_width: 1280
      detail: low
      max_tokens: 100
      temperature: 0.1
      expose_images: true
      provider: 01JEKQS47AA4SACQHXHF3J18VP
      image_entity:
        - camera.garasje2
      message: >
        Analyze the camera image and return a JSON object only, without
        additional quotation marks or formatting:

        {
          "trigger_open": true/false,
          "debug_message": "A descriptive message for debugging"
        }

        - "trigger_open": true if the following criteria are met:
            1. A dark gray car is identified as a Tesla.
            2. The license plate reads "ABC 123456".
            3. The car is reversing and is close to two white poles.
            4. Movement between frames confirms the car is reversing.

        - "trigger_open": false if no cars are identified or if the criteria are
        not met.

        Only return the JSON object exactly as shown above. Do not include
        additional text, quotation marks, or formatting around the JSON object.
    response_variable: response
  - choose:
      - conditions:
          - condition: template
            value_template: >-
              {% set response = response.response_text | from_json %} {{
              response.trigger_open == true }}
          - condition: state
            entity_id: sensor.garage_state
            state: Vehicle away,Garage closed
        sequence:
          - data:
              message: "{{ response.response_text }}"
            action: notify.mobile_app_cell_phone
          - device_id: 4c257e0a8a3cbb38fce126b49fbc5bd0
            domain: cover
            entity_id: fa81b026593cd88d40fcebc4cc095ad3
            type: open
  - action: notify.mobile_app_cell_phone
    data:
      message: >-
        Ingen match: {{ (response.response_text | from_json).trigger_open }}, 
        Message: {{ (response.response_text | from_json).debug_message }}
`

**Comment by valentinfrlch** [comment]
There is a potential bug currently (#85) that mixes up the order of the images.
Did you run into any issues with your automation?

**Comment by torsteinelv** [comment]
Thank you for the reply.

Yes, while I was testing the automation, the car has been parked still in the driveway, but the response from the LLM indicates that the car is moving and reversing towards the garage.

So, it works almost perfectly. The only issue I encountered is that I want it to avoid detection if the car is driving forward or standing still.

I have looked at their blueprint and tested it. It manages to understand when I drive in and out again afterward. However, I haven’t been able to achieve this in my own example here.

I also made an attempt to fetch the R state directly from the car, but unfortunately, it’s too slow for the automation to capture this in time.

Perhaps someone here has tips for the final touches, as I think this automation can work well.

**Comment by valentinfrlch** [comment]
Difficult to tell. It might not be the chronological order but instead might just be the typical "AI hallucination". Personally, I would try to use the cars sensor directly as AI might also mistake another car for yours. You mentioned your car is a Tesla. Don't Teslas have "HomeLink"? Maybe there is a way to emulate such a HomeLink with Home Assistant.

**Comment by torsteinelv** [comment]
I can already read all the sensors from the car in Home Assistant, so it works, but it's more interesting to understand how the LLMVision integration you've created for Home Assistant works and how it can be used.

I’m already checking license plates and describing the car and understand that LLMs can sometimes hallucinate.

Previously, I’ve also noticed that LLMVision can describe differences between various frames captured, so I might experiment a bit more with prompts to see what I can achieve.

I’ll try to experience with different prompts to see if I can get it to work better 😊

**Comment by torsteinelv** [comment]
Will do. I tried to add the stream of pictures to the notification as extra debugging. Anyone knows how this can be done?

**Comment by rsteckler** [comment]
You can add a parameter ```expose_images: true``` and it will save the frames in /config/www/llmvision.

**Comment by torsteinelv** [comment]
![Screenshot_20241210_161656_Home Assistant](https://github.com/user-attachments/assets/96a70704-8a28-43b7-a8e6-da4903df56ef)
Tried the description. Still dont work while car is parked and another car drives by 😊

**Comment by rsteckler** [comment]
If you look at the first/last images in your /config/www/llmvision folder, are they in the correct order and do they show the car, etc?

**Comment by torsteinelv** [comment]
Been testing some more, this prompt works pretty good, so I have been testing with video analyzer since it works with not live clips. mabey that also is a solution in order to make this work better. Also added close garage functionality :)

Will try to leave it with notifications only for a long time and see how many false opening / closes I get.


`Analyze at least 8 consecutive frames and return a valid JSON object only after evaluating all frames. Do not include any text, formatting, or code blocks (e.g., no backticks ` or ``` around the JSON). Return only the JSON object exactly in the format below:

{
  "trigger_open": true/false,
  "trigger_close": true/false,
  "debug_message": "A descriptive message for debugging, including a frame-by-frame explanation."
}

General Logic:
1. Evaluate the car’s type and license plate:
   - The car must be a dark grey Tesla Model Y with license plate "AB 1234".
   - If this car is not detected, both "trigger_open" and "trigger_close" must be false.

Conditions to set "trigger_open" to true:
1. The car consistently moves closer to the camera across at least 3 consecutive frames, exceeding a total movement threshold of 30 pixels.
2. The car stops close to or between the two poles.
3. The car **must be visible in all frames** during the sequence.
4. If these conditions are met, "trigger_close" must automatically be set to false.

Conditions to set "trigger_close" to true:
1. A person is detected entering the car.
2. The car consistently moves away from the camera across at least 3 consecutive frames, exceeding a total movement threshold of 30 pixels.
3. The car **must be visible in all frames** during the sequence.
4. If these conditions are met, "trigger_open" must automatically be set to false.

If neither condition is fully met:
1. Both "trigger_open" and "trigger_close" must be false.
2. The debug message must include:
   - Frame-by-frame analysis of car movement.
   - Information about detected persons near or entering the car.
   - A conclusion explaining why no action was triggered.

Additional Debugging:
1. If the car disappears in any frame, explain this as an error and set both "trigger_open" and "trigger_close" to false.
2. Log the exact pixel movement (absolute values) of the car between frames for transparency.
`

**Comment by torsteinelv** [comment]
you can
